src/advanced.sml

```diff
@@ -0,0 +1,416 @@
+       | NONE => false
+
+  (* Discard useless card in our hand (if possible). *)
+  fun discardUseless (s : state) : action option =
+    if #hints s = 8
+    then NONE
+    else Option.map Discard (ourOldestUseless s (#clues s))
+
+  (* Discard oldest unclued card in our hand (if possible). *)
+  fun discardOldestUnclued (s : state) : action option =
+    if #hints s = 8
+    then NONE
+    else Option.map Discard (oldestUnclued (#clues s))
+
+  (* Try to give a hint pointing at 0 cards.
+   * Currently it will only try to give rank hints to the next player, but this can be extended *)
+  fun giveBlankHint (s : state) : action option =
+    if #hints s = 0
+    then NONE
+    else Option.map (fn i => HintRank (0, i))
+		    (List.find (fn i => not (List.exists (fn x => #2 (#1 x) = i) (hd (#hands s))))
+			       [5,4,3,2,1])
+
+  (* Give a hint which cannot be interpreted as a play hint
+   * This will hint 5 if that hint cannot be interpreted as a play hint;
+   * Otherwise it will hint 1 if all 1s have been played;
+   * Otherwise it will try to give a blank hint *)
+  fun wasteHint (s : state) : action option =
+    if #hints s = 0
+    then NONE
+    else if List.all (fn su => SD.lookup (#inPlay s) su <> 4) (SSet.toList suits)
+    then SOME (HintRank (0, 5))
+    else if List.all (fn su => SD.lookup (#inPlay s) su > 0) (SSet.toList suits)
+    then SOME (HintRank (0, 1))
+    else giveBlankHint s
+
+  val otherwise = Util.otherwise
+  infix 4 otherwise
+
+  (* Try to find the least harmful discard action.
+   * It first tries to discard the oldest card known to be not vital (the oldest);
+   * Then it will try to discard a card not marked with number (the newest);
+   * (TO DO: discard card which was never saved when it was on the chop)
+   * Otherwise it will discard the card clued with the highest rank (the oldest).
+   * This function assumes that discardOldestUnclued returned NONE, and only returns NONE if there
+     are 8 hints available *)
+  fun discardAny (s : state) : action option =
+    if #hints s = 8 then NONE else
+    SOME (Option.map Discard (Util.revFindIndex (isNotVital s) (#clues s)) otherwise (fn () =>
+          Option.map Discard (Util.findIndex (not o isSome o cluedRank) (#clues s))
+                     otherwise (fn () =>
+          Discard (Util.revFindMaxIndex (valOf o cluedRank) (#clues s)))))
+
+  (* The order of actions should become:
+   * - (respond to finesse)
+   * - if chop of next player is vital or playable:
+   *   = give play clue on a card which is not clued in your hand to next player
+   *   = give a save clue
+   * - play a playable card
+   * - give any play clue *)
+  fun play (num : int) : state -> action =
+  let
+    val m : memory ref = ref (initialMemory num)
+  in
+    fn s => (
+    m := withPlayable (!m) (updatePlayable s (#playable (!m)));
+    (* print (memoryToString (!m) ^"\n"); *)
+    (case playPlayableCard (hd (#playable (!m))) of
```

Comment:
That's indeed nicer. It's now
```
   Option.map (fn i => giveSaveHint s otherwise (fn () => Play i))
               (Util.findIndex (fn b => b) (hd (#playable (!m))))
```